### PR TITLE
optional postgress database #322

### PR DESCRIPTION
**Pull request by steve7158**
Steps to implement this:
1. open tanner/data/psql_seup.py, change <pwd>  to your default postgresql user('postgres) password.
2. Run `sudo python3 tanner/data/psql_setup.py`
3. Make the same change in password in tanner/config.py in dictionary config_template under the key 'POSTGRES'
4. Run `sudo python3 setup.py install`
5. Run `sudo tanner --database postgres`

### CODE REVIEW

File: tanner/data/psql_setup.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+from sqlalchemy import create_engine
+engine=create_engine("postgresql://postgres:<pwd>@localhost/postgres")
+conn=engine.connect()
+conn.execute("commit")
+conn.execute("create database tanner_db")
```
Comment (glaslos) [inline tanner/data/psql_setup.py]: on tanner startup, run `CREATE TABLE IF NOT EXISTS tanner` (also rename the db to just `tanner`)

File: tanner/server.py [hunk 2]
```diff
@@ -20,7 +20,7 @@
 
 
 class TannerServer:
-    def __init__(self):
+    def __init__(self,database):
```
Comment (glaslos) [inline tanner/server.py]: pep8

File: tanner/server.py [hunk 3]
```diff
@@ -30,7 +30,8 @@ def __init__(self):
         self.dorks = dorks_manager.DorksManager()
         self.base_handler = base.BaseHandler(base_dir, db_name)
         self.logger = logging.getLogger(__name__)
-        self.redis_client = None
+        self.db_client = None
+        self.database=database
```
Comment (glaslos) [inline tanner/server.py]: pep8 (configure your IDE to fix formatting issues)

File: tanner/data/psql_setup.py (hunk 1 above)
Comment (steve7158) [reply tanner/data/psql_setup.py]: what should i add for columns?

Comment (steve7158) [reply tanner/data/psql_setup.py]: what should i add for columns?

Comment (glaslos) [reply tanner/data/psql_setup.py]: you mean creating the tables?
We want to replicate the same functionality we have with Redis. Have a look at the session_manager and dorks.

**Comment by glaslos** [comment]
Unless you decided to stop working on this issue, please keep iterating in the same PR so we maintain the history.

**Comment by steve7158** [comment]
@glaslos Thank you for informing me ill reopen this PR

**Comment by afeena** [comment]
@steve7158 please let us know when you are ready to get some feedback :)

**Comment by steve7158** [comment]
Thank you @afeena, Sure i will definitely let you guys know about it, sorry the delay, the exams are near so its difficult to give too much time for developing.

**Comment by glaslos** [comment]
@steve7158 no worries, the request wasn't about us impatiently waiting for you to finish, but us seeing your commits and wondering at which point you would like to get feedback :)
Good luck with your exams, they obviously have priority!

**Comment by steve7158** [comment]
Thank you @glaslos

**Comment by steve7158** [comment]
@afeena @glaslos can you please give me some feedback for the status so far.
I made 2 tables, because of zadd in session analyzer, tell me is it necessary to add in that manner cant it be a something simple like sadd.

**Comment by afeena** [comment]
@glaslos Do we want to completely move from redis to postgresql? Or we suppose to have a possibility to choose? Or probably even melt everything together (use redis as a  memory and postgresql as storage)?

**Comment by glaslos** [comment]
@afeena I'd suggest to go with postgres only for the beginning and introduce a redis cache later.
@steve7158 you definitely want to go away from using postgres like a key value store. Use the relational feature, normalize data... Let me know if you need help with the data model.

**Comment by steve7158** [comment]
Thank you @glaslos , but what is relational feature and normalize data. Do you have some documentation or tutorial where i can study from? :sweat_smile: :smile:

**Comment by steve7158** [comment]
and yes please tell me how the data model should be like. :smile:

File: bin/tanner [hunk 4]
```diff
@@ -17,7 +17,9 @@ def main():
     """)
     parser = argparse.ArgumentParser()
     parser.add_argument("--config", help="tanner config")
+    parser.add_argument("--database", help="choose a database postgres/redis", default='redis')
```
Comment (afeena) [inline bin/tanner]: Since we agree with @glaslos that we should use only postgresql for now, it's not necessary to set db as a parameter

Comment (steve7158) [reply bin/tanner]: Ok :+1: ill remove that.

**Comment by glaslos** [comment]
Python + Postgres: https://www.postgresqltutorial.com/postgresql-python/
DB normalization: https://en.wikipedia.org/wiki/Database_normalization
@afeena do you have sample data of what we store for the sessions now? Or could you give an intro on how we could model the data in Postgres?

**Comment by afeena** [comment]
Snare data example:
```
{
  "sess_uuid": "e166441850e2464da678dcc070799e89",
  "peer_ip": "88.77.0.131",
  "peer_port": 50363,
  "location": {
    "country": "Germany",
    "country_code": "DE",
    "city": "Duisburg",
    "zip_code": "47137"
  },
  "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML,like Gecko) Chrome/80.0.3987.163 Safari/537.36",
  "snare_uuid": "776edef4-ec5f-4fe0-8485-02cf9ff1f64d",
  "start_time": 1586510182.2908487,
  "end_time": 1586510182.2908504,
  "requests_in_second": 599186.2857142857,
  "approx_time_between_requests": 0,
  "accepted_paths": 1,
  "errors": 0,
  "hidden_links": 0,
  "attack_types": [
    "index"
  ],
  "attack_count": {
    "index": 1
  },
  "paths": [
    {
      "path": "/",
      "timestamp": 1586510182.2907968,
      "response_status": 200,
      "attack_type": "index"
    }
  ],
  "cookies": {
    "_ga": "GA1.1.1415803245.1583793692",
    " ga-disable-108255672-1": "true",
    " sess_uuid": "01b8be92-ecc6-49d3-a0dc-df9510408fd4",
    "sess_uuid": null
  },
  "referer": null,
  "possible_owners": {
    "user": 0.15,
    "tool": 0.85,
    "crawler": 0.5
  }
}

```

Session data  example:
```
{
  "peer": {
    "ip": "84.56.155.136",
    "port": 48194
  },
  "user_agent": "Mozilla/5.0 (X11; Fedora; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.122 Safari/537.36",
  "snare_uuid": "776edef4-ec5f-4fe0-8485-02cf9ff1f64d",
  "sess_uuid": "3d09dbb58465444684cbfb9b7d8a313f",
  "start_time": 1583352304.5080535,
  "end_time": 1583352304.508056,
  "count": 1,
  "paths": [
    {
      "path": "/",
      "timestamp": 1583352304.50796,
      "response_status": 200,
      "attack_type": "index"
    }
  ],
  "cookies": {
    "sess_uuid": null
  },
  "referer": null
}
```

**Comment by afeena** [comment]
@steve7158 What is the status of this pull request? :)

**Comment by steve7158** [comment]
> @steve7158 What is the status of this pull request? :)

I haven't made any updates since my last PR.

**Comment by afeena** [comment]
@steve7158 we are going to work on this issue as a part of GSoC 2020, would you like to collaborate in some way? Are you willing to continue this work?

**Comment by steve7158** [comment]
> @steve7158 we are going to work on this issue as a part of GSoC 2020, would you like to collaborate in some way? Are you willing to continue this work?

@afeena, I wont be working on this issue however I will be collaborating by fixing bugs if possible. 😄  Best  of luck for GSOC 2020 👍

**Comment by afeena** [comment]
@steve7158 then I am closing this PR, feel free to create new one when you have time or willing to work on other problem :)
Thanks and good luck you too ;)

**Pull request closed without merging**